### DISCUSSION TITLE
Auto-generate color shades for custom colors defined in tailwind.config.js.

### DISCUSSION THREAD

**Discussion by LorenzoBloedow** [tailwindlabs/tailwindcss#9654] [Ideas]
This is a PR I want to make so that colors defined both under `module.exports.colors` and `module.exports.extend.colors` get automatically generated shades akin to the one we already have with built-in colors like `bg-slate-900`.
I just feel like Tailwind is really missing this feature and I would love to start coding it ASAP, so I'm starting this discussion to know if I missed it and this is already being developed and I would also like to know your thoughts on it!

**Comment by cjoecker** [comment]
I like this idea! I would create a flag in the colors, to autogenerate the colors if it is true. Something like this:
```
module.exports = {
  theme: {
    extend: {
      colors: {
        autogenerateShades: true
        blue: '#17275c'
      }
    },
  },
}
```

I would allow as value a `boolean` to create 10 shades by default, but also `number` to autogenerate a specific number of shades.
If the developer writes `3`, then only 3 different shades are created. This may be useful in case you want to create a `dark`, `light`, and `DEFAULT` tone.

Another option is to have a flag named `autogenerate` that accepts two values: `shades` or `lightAndDark`. If the value is `shades`, then 10 numbered shades are created. If `lightAndDark` is selected, then a `DEFAULT`, `-dark` and `-light` color is created.

**Reply by LorenzoBloedow** [reply]
Thank you for the feedback! Definitely going to implement that flag, I like the idea of having it accept either a boolean or a number. Should the default value be true?

**Reply by cjoecker** [reply]
🤔 good question. I would not make it default since I guess that the majority of the developers won't use it.
The next question is, if you have for example `3`, how separated the shades should be? 33% 50% and 66%? That's why I was thinking about the option of `shades` or `lightAndDark`

**Reply by LorenzoBloedow** [reply]
I was thinking it could get split into 10 different shades (because 10 would be the maximum value, which you would set with either `true` or `10`) with each shade being 10% in the 'lightness' value of HSL and it would start to distribute it in the middle always first going down and then up, so for instance: `autoGenerateShades: 3` would be 50% -> 40% -> 60% and `autoGenerateShades: 4` would be 50% -> 40% -> 60% -> 30%

**Reply by cjoecker** [reply]
Aa ok, understood. I'm wondering in which cases you would use for example `3`. 
In my experience is because I just want to differentiate between three tones: default, dark and light. If the distance is only 10%, it might be too low the differentiate them. Therefore, I was thinking of making the shade distance bigger, the lower the number of `autogenerate` value is.

**Reply by LorenzoBloedow** [reply]
You're right that's a much better way of doing it, so something like: `10 / (autoGenerateShadesValue * 10) * the color's current lightness value`? For example `autoGenerateShades: 2` on `hsl(0, 100%, 26%)` would be: `10 / 20 = 0.5` `0.5 * 26 = 13` and then the shades' values would be 13% down and 13% up, so `26 - 13 =` `hsl(0, 100%, 13%)` and `26 + 13 =` `hsl(0, 100%, 39%)`.

**Reply by cjoecker** [reply]
Yeah, I like that! 
But now that we are having this discussion, I think it is better to only two options that are the most common: `shades` or `lightAndDark`. Then it is easier to guess what Tailwind does. 
But as you wish, both options are good for me :)

**Reply by LorenzoBloedow** [reply]
I still think it's better to go with either a number or a boolean but I'm hoping to get a few more opinions on this, so if anyone else is reading this, please do tell me what you think! :)

**Reply by LorenzoBloedow** [reply]
An update on this: I ended up doing it as a plugin!
https://www.npmjs.com/package/tailwind-shades-for-custom-colors

**Reply by cjoecker** [reply]
Very nice, thanks!!!
